File: specs/typespec/hub/endorsements.ts
import type { Handle } from "./hub-users.js";
import type { ConnectionState } from "./connections.js";

export type EndorsementRequestState =
	| "pending"
	| "written"
	| "declined"
	| "expired";

export interface RequestEndorsementsRequest {
	application_id: string;
	endorser_handles: Handle[];
	note?: string;
}

export interface EndorsementRequestIncoming {
	request_id: string;
	application_id: string;
	candidate_handle: Handle;
	candidate_display_name: string;
	org_domain: string;
	org_name: string;
	opening_title: string;
	shared_domain: string;
	overlap_start_year: number;
	overlap_end_year: number;
	note?: string;
	state: EndorsementRequestState;
	requested_at: string;
	candidate_connection_state: ConnectionState;
}

export interface EndorsementRequestOutgoing {
	request_id: string;
	application_id: string;
	endorser_handle: Handle;
	endorser_display_name: string;
	state: EndorsementRequestState;
	requested_at: string;
}

export interface ListEndorsementRequestsIncomingRequest {
	filter_state?: EndorsementRequestState[];
	pagination_key?: string;
	limit?: number;
}

export interface ListEndorsementRequestsIncomingResponse {
	requests: EndorsementRequestIncoming[];
	next_pagination_key?: string;
}

export interface ListEndorsementRequestsOutgoingRequest {
	application_id: string;
	pagination_key?: string;
	limit?: number;
}

export interface ListEndorsementRequestsOutgoingResponse {
	requests: EndorsementRequestOutgoing[];
	next_pagination_key?: string;
}

export interface WriteEndorsementRequest {
	request_id?: string;
	application_id?: string;
	text: string;
}

export interface WriteEndorsementResponse {
	endorsement_id: string;
}

export interface UpdateEndorsementRequest {
	endorsement_id: string;
	text: string;
}

export interface DeclineEndorsementRequestRequest {
	request_id: string;
}

export interface HideEndorsementOnApplicationRequest {
	endorsement_id: string;
}

export interface ShowEndorsementOnApplicationRequest {
	endorsement_id: string;
}

export interface ValidationError {
	field: string;
	message: string;
}

export function validateRequestEndorsementsRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.application_id !== "string" || r.application_id.trim() === "") {
		errors.push({
			field: "application_id",
			message: "Must be a non-empty string",
		});
	}

	if (!Array.isArray(r.endorser_handles)) {
		errors.push({ field: "endorser_handles", message: "Must be an array" });
	} else {
		if (r.endorser_handles.length < 1 || r.endorser_handles.length > 10) {
			errors.push({
				field: "endorser_handles",
				message: "Must have 1-10 items",
			});
		}
		for (const h of r.endorser_handles) {
			if (typeof h !== "string" || h.trim() === "") {
				errors.push({
					field: "endorser_handles",
					message: "All handles must be non-empty strings",
				});
				break;
			}
		}
	}

	if (r.note !== undefined) {
		if (typeof r.note !== "string") {
			errors.push({ field: "note", message: "Must be a string" });
		} else if (r.note.length > 500) {
			errors.push({
				field: "note",
				message: "Must be at most 500 characters",
			});
		}
	}

	return errors;
}

export function validateListEndorsementRequestsIncomingRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (r.limit !== undefined) {
		if (typeof r.limit !== "number" || r.limit < 1 || r.limit > 100) {
			errors.push({ field: "limit", message: "Must be between 1 and 100" });
		}
	}

	if (r.pagination_key !== undefined && typeof r.pagination_key !== "string") {
		errors.push({ field: "pagination_key", message: "Must be a string" });
	}

	return errors;
}

export function validateListEndorsementRequestsOutgoingRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.application_id !== "string" || r.application_id.trim() === "") {
		errors.push({
			field: "application_id",
			message: "Must be a non-empty string",
		});
	}

	if (r.limit !== undefined) {
		if (typeof r.limit !== "number" || r.limit < 1 || r.limit > 100) {
			errors.push({ field: "limit", message: "Must be between 1 and 100" });
		}
	}

	return errors;
}

export function validateWriteEndorsementRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	const hasRequestId =
		typeof r.request_id === "string" && r.request_id.trim() !== "";
	const hasApplicationId =
		typeof r.application_id === "string" && r.application_id.trim() !== "";

	if (hasRequestId === hasApplicationId) {
		errors.push({
			field: "request_id",
			message: "Exactly one of request_id or application_id must be provided",
		});
	}

	if (typeof r.text !== "string") {
		errors.push({ field: "text", message: "Must be a string" });
	} else if (r.text.length < 100 || r.text.length > 2000) {
		errors.push({
			field: "text",
			message: "Must be between 100 and 2000 characters",
		});
	}

	return errors;
}

export function validateUpdateEndorsementRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.endorsement_id !== "string" || r.endorsement_id.trim() === "") {
		errors.push({
			field: "endorsement_id",
			message: "Must be a non-empty string",
		});
	}

	if (typeof r.text !== "string") {
		errors.push({ field: "text", message: "Must be a string" });
	} else if (r.text.length < 100 || r.text.length > 2000) {
		errors.push({
			field: "text",
			message: "Must be between 100 and 2000 characters",
		});
	}

	return errors;
}

export function validateDeclineEndorsementRequestRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.request_id !== "string" || r.request_id.trim() === "") {
		errors.push({ field: "request_id", message: "Must be a non-empty string" });
	}

	return errors;
}

export function validateHideEndorsementOnApplicationRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.endorsement_id !== "string" || r.endorsement_id.trim() === "") {
		errors.push({
			field: "endorsement_id",
			message: "Must be a non-empty string",
		});
	}

	return errors;
}

export function validateShowEndorsementOnApplicationRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.endorsement_id !== "string" || r.endorsement_id.trim() === "") {
		errors.push({
			field: "endorsement_id",
			message: "Must be a non-empty string",
		});
	}

	return errors;
}

File: api-schema/hub/endorsements.go
package hub

import (
	"vetchium-api-server.typespec/common"
)

type EndorsementRequestState string

const (
	EndorsementRequestStatePending  EndorsementRequestState = "pending"
	EndorsementRequestStateWritten  EndorsementRequestState = "written"
	EndorsementRequestStateDeclined EndorsementRequestState = "declined"
	EndorsementRequestStateExpired  EndorsementRequestState = "expired"
)

type RequestEndorsementsRequest struct {
	ApplicationID   string   `json:"application_id"`
	EndorserHandles []string `json:"endorser_handles"`
	Note            *string  `json:"note,omitempty"`
}

type EndorsementRequestIncoming struct {
	RequestID                string                  `json:"request_id"`
	ApplicationID            string                  `json:"application_id"`
	CandidateHandle          string                  `json:"candidate_handle"`
	CandidateDisplayName     string                  `json:"candidate_display_name"`
	OrgDomain                string                  `json:"org_domain"`
	OrgName                  string                  `json:"org_name"`
	OpeningTitle             string                  `json:"opening_title"`
	SharedDomain             string                  `json:"shared_domain"`
	OverlapStartYear         int32                   `json:"overlap_start_year"`
	OverlapEndYear           int32                   `json:"overlap_end_year"`
	Note                     *string                 `json:"note,omitempty"`
	State                    EndorsementRequestState `json:"state"`
	RequestedAt              string                  `json:"requested_at"`
	CandidateConnectionState string                  `json:"candidate_connection_state"`
}

type EndorsementRequestOutgoing struct {
	RequestID           string                  `json:"request_id"`
	ApplicationID       string                  `json:"application_id"`
	EndorserHandle      string                  `json:"endorser_handle"`
	EndorserDisplayName string                  `json:"endorser_display_name"`
	State               EndorsementRequestState `json:"state"`
	RequestedAt         string                  `json:"requested_at"`
}

type ListEndorsementRequestsIncomingRequest struct {
	FilterState   []EndorsementRequestState `json:"filter_state,omitempty"`
	PaginationKey *string                   `json:"pagination_key,omitempty"`
	Limit         *int32                    `json:"limit,omitempty"`
}

type ListEndorsementRequestsIncomingResponse struct {
	Requests          []EndorsementRequestIncoming `json:"requests"`
	NextPaginationKey *string                      `json:"next_pagination_key,omitempty"`
}

type ListEndorsementRequestsOutgoingRequest struct {
	ApplicationID string  `json:"application_id"`
	PaginationKey *string `json:"pagination_key,omitempty"`
	Limit         *int32  `json:"limit,omitempty"`
}

type ListEndorsementRequestsOutgoingResponse struct {
	Requests          []EndorsementRequestOutgoing `json:"requests"`
	NextPaginationKey *string                      `json:"next_pagination_key,omitempty"`
}

type WriteEndorsementRequest struct {
	RequestID     *string `json:"request_id,omitempty"`
	ApplicationID *string `json:"application_id,omitempty"`
	Text          string  `json:"text"`
}

type WriteEndorsementResponse struct {
	EndorsementID string `json:"endorsement_id"`
}

type UpdateEndorsementRequest struct {
	EndorsementID string `json:"endorsement_id"`
	Text          string `json:"text"`
}

type DeclineEndorsementRequestRequest struct {
	RequestID string `json:"request_id"`
}

type HideEndorsementOnApplicationRequest struct {
	EndorsementID string `json:"endorsement_id"`
}

type ShowEndorsementOnApplicationRequest struct {
	EndorsementID string `json:"endorsement_id"`
}

func (r RequestEndorsementsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.ApplicationID == "" {
		errs = append(errs, common.ValidationError{Field: "application_id", Message: "Must be a non-empty string"})
	}
	if len(r.EndorserHandles) < 1 || len(r.EndorserHandles) > 10 {
		errs = append(errs, common.ValidationError{Field: "endorser_handles", Message: "Must have 1-10 items"})
	}
	for _, h := range r.EndorserHandles {
		if h == "" {
			errs = append(errs, common.ValidationError{Field: "endorser_handles", Message: "All handles must be non-empty strings"})
			break
		}
	}
	if r.Note != nil && len(*r.Note) > 500 {
		errs = append(errs, common.ValidationError{Field: "note", Message: "Must be at most 500 characters"})
	}
	return errs
}

func (r ListEndorsementRequestsIncomingRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > 100) {
		errs = append(errs, common.ValidationError{Field: "limit", Message: "Must be between 1 and 100"})
	}
	return errs
}

func (r ListEndorsementRequestsOutgoingRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.ApplicationID == "" {
		errs = append(errs, common.ValidationError{Field: "application_id", Message: "Must be a non-empty string"})
	}
	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > 100) {
		errs = append(errs, common.ValidationError{Field: "limit", Message: "Must be between 1 and 100"})
	}
	return errs
}

func (r WriteEndorsementRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	hasReqID := r.RequestID != nil && *r.RequestID != ""
	hasAppID := r.ApplicationID != nil && *r.ApplicationID != ""
	if hasReqID == hasAppID {
		errs = append(errs, common.ValidationError{Field: "request_id", Message: "Exactly one of request_id or application_id must be provided"})
	}
	if len(r.Text) < 100 || len(r.Text) > 2000 {
		errs = append(errs, common.ValidationError{Field: "text", Message: "Must be between 100 and 2000 characters"})
	}
	return errs
}

func (r UpdateEndorsementRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.EndorsementID == "" {
		errs = append(errs, common.ValidationError{Field: "endorsement_id", Message: "Must be a non-empty string"})
	}
	if len(r.Text) < 100 || len(r.Text) > 2000 {
		errs = append(errs, common.ValidationError{Field: "text", Message: "Must be between 100 and 2000 characters"})
	}
	return errs
}

func (r DeclineEndorsementRequestRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.RequestID == "" {
		errs = append(errs, common.ValidationError{Field: "request_id", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r HideEndorsementOnApplicationRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.EndorsementID == "" {
		errs = append(errs, common.ValidationError{Field: "endorsement_id", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r ShowEndorsementOnApplicationRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.EndorsementID == "" {
		errs = append(errs, common.ValidationError{Field: "endorsement_id", Message: "Must be a non-empty string"})
	}
	return errs
}
